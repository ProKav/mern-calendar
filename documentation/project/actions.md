<!-- prettier-ignore -->
### Actions

---

Summary of the actions that each function executes within the code

```javascript
export const startLogin = (email, password) => {

}
```

| Actions                                                    |
| ---------------------------------------------------------- |
| Send email and password                                    |
| Connect with fetchWithOutToken throught GET method and url |
| Save token in localStorage                                 |
| Save date when token is saved                              |
| Dispatch login or error                                    |

---

```javascript
export const startRegister = (name, email, password) => {
	
}
```

| Actions |
| ------- |
| Send name, email and password        |
| Connect with fetchWithOutToken throught POST method and url        |
| Save token in localStorage        |
| Save date when token is saved        |
| Dispatch login or error        |

---

```javascript
export const startChecking = () => {
	
}
```

| Actions |
| ------- |
| Send token to fetchWithToken for renew it        |
| Save token in localStorage        |
| Save date when token is saved        |
| Dispatch login or checkingFinish        |

---

```javascript
export const startLogout = () => {
	
}
```

| Actions |
| ------- |
| Clear localStorage        |
| Clear active event        |
| Logout        |

### [Back](../README.md) :leftwards_arrow_with_hook:
