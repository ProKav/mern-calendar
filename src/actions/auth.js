import { fetchWithoutToken } from '../helpers/fetchWithoutToken'
import { fetchWithToken } from '../helpers/fetchWithToken'
import { types } from '../types/types'
import Swal from 'sweetalert2'

export const startLogin = (email, password) => {

	return async dispatch => {
		const resp = await fetchWithoutToken('auth', { email, password }, 'POST')
		const body = await resp.json()
		const error_msg = body.errors ? Object.values(body.errors)[0].msg : body.msg

		if (body.ok) {
			localStorage.setItem('token', body.token)
			localStorage.setItem('token-init-date', new Date().getTime())

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				}),
			)
		} else {
			Swal.fire('Error', error_msg, 'error')
		}
	}
}

export const startRegister = (name, email, password) => {
	return async dispatch => {
		const resp = await fetchWithoutToken('auth/new', { name, email, password }, 'POST')
		const body = await resp.json()
		const error_msg = body.errors ? Object.values(body.errors)[0].msg : body.msg
		if (body.ok) {
			localStorage.setItem('token', body.token)
			localStorage.setItem('token-init-date', new Date().getTime())

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				}),
			)
		} else {
			Swal.fire('Error', error_msg, 'error')
		}
	}
}

export const startChecking = () => {
	return async (dispatch) => {
		const resp = await fetchWithToken('auth/renew')
		const body = await resp.json()

		if (body.ok) {
			localStorage.setItem('token', body.token)
			localStorage.setItem('token-init-date', new Date().getTime())

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				}),
			)
		} else {
			dispatch(checkingFinish())
		}
	}
}

const checkingFinish = () => ({
	type: types.authCheckingFinish,
})

const login = user => ({
	type: types.authLogin,
	payload: user,
})
