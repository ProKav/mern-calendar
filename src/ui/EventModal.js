import React, { useState } from 'react'
import DateTimePicker from 'react-datetime-picker';
import moment from 'moment'

const now = moment().minutes(0).seconds(0).add(1, 'hours')
const end = now.clone().add(1, 'hours')

export const EventModal = () => {

    const [dateStart, setDateStart] = useState(now.toDate());
    const [dateEnd, setDateEnd] = useState(end.toDate());

    const handleStartDateChange = (e) => {
        setDateStart(e)
        console.log(e)
    }

    const handleEndDateChange = (e) => {
        setDateEnd(e)
    }

    return (
        <>
            <h1 className='__modal-h1 mt-1'> Nuevo evento </h1>
            <hr />
            <form className="container">

                <div className="form-group">
                    <label className='mt-2'>Fecha y hora inicio</label>
                    <DateTimePicker
                        onChange={handleStartDateChange}
                        value={dateStart}
                        className='form-control'
                    />
                </div>

                <div className="form-group">
                    <label className='mt-2'>Fecha y hora fin</label>
                    <DateTimePicker
                        onChange={handleEndDateChange}
                        value={dateEnd}
                        minDate={ dateStart }
                        className='form-control'
                    />
                </div>

                <hr />
                <div className="form-group">
                    <label className='mt-2'>Titulo y notas</label>
                    <input
                        type="text"
                        className="form-control"
                        placeholder="Título del evento"
                        name="title"
                        autoComplete="off"
                    />
                    <small id="emailHelp" className="form-text text-muted">Una descripción corta</small>
                </div>

                <div className="form-group">
					<textarea
                        type="text"
                        className="form-control"
                        placeholder="Notas"
                        rows="5"
                        name="notes"
                    ></textarea>
                    <small id="emailHelp" className="form-text text-muted">Información adicional</small>
                </div>

                <button
                    type="submit"
                    className="btn btn-outline-primary btn-block"
                >
                    <i className="far fa-save"></i>
                    <span> Guardar</span>
                </button>

            </form>
        </>
    )
}