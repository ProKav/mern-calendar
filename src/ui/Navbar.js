import React from 'react'
import { useSelector } from 'react-redux';
import { hello } from '../helpers/hello';

export const Navbar = () => {
const {name} = useSelector(state => state.auth)
  
	return (
		<div className='navbar navbar-dark bg-dark'>
			<span className='navbar-brand'>{`Hola! ${hello()}${name}`}</span>
			<button className='btn btn-outline-danger'>
				<i className='fas fa-sign-out-alt'></i>
				<span> Salir</span>
			</button>
		</div>
	)
}
