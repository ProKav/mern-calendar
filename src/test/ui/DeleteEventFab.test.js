import React from 'react'
import { mount } from 'enzyme'
import { DeleteEventFab } from '../../ui/DeleteEventFab'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import '@testing-library/jest-dom'
import { eventStartDelete } from '../../actions/events'

jest.mock('../../actions/events', () => ({
	eventStartDelete: jest.fn(),
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {}
const store = mockStore(initState)
store.dispatch = jest.fn()

const wrapper = mount(
	<Provider store={store}>
		<DeleteEventFab />
	</Provider>,
)

describe('Test in <DeleteEventFab />', () => {
	test('Should be displayed correctly', () => {
		expect(wrapper).toMatchSnapshot()
	})

	test('Should call eventStartDelete when clicked', () => {
		wrapper.find('button').prop('onClick')()

		expect(eventStartDelete).toHaveBeenCalled()
	})
})
