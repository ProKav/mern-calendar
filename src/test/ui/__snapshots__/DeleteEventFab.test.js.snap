// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <DeleteEventFab /> Should be displayed correctly 1`] = `
<button
  className="btn btn-danger __ui-fab-delete"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Borrar Evento
  </span>
</button>
`;
